packages/dialog/src/lit/renderer-directives.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license
+ * Copyright (c) 2017 - 2022 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+/* eslint-disable max-classes-per-file */
+import { directive } from 'lit/directive.js';
+import { microTask } from '@vaadin/component-base/src/async.js';
+import { Debouncer } from '@vaadin/component-base/src/debounce.js';
+import { LitRendererDirective } from '@vaadin/lit-renderer';
+
+const CONTENT_UPDATE_DEBOUNCER = Symbol('contentUpdateDebouncer');
+
+class AbstractDialogRendererDirective extends LitRendererDirective {
+  /**
+   * A property to that the renderer callback will be assigned.
+   *
+   * @abstract
+   */
+  rendererProperty;
+
+  /**
+   * Adds the renderer callback to the dialog.
+   */
+  addRenderer() {
+    this.element[this.rendererProperty] = (root, dialog) => {
+      this.renderRenderer(root, dialog);
+    };
+  }
+
+  /**
+   * Runs the renderer callback on the dialog.
+   */
+  runRenderer() {
+    this.element[CONTENT_UPDATE_DEBOUNCER] = Debouncer.debounce(
```

Comment:
Is this debouncer going to be added to all renderer implementations? In that case, should it be moved to the abstract `LitRendererDirective`?

Reply:
> Is the debouncing necessary in general (LitElement already batches render)?

Some sort of batching is good to have because when you cause Lit to update several directives, for example by updating their dependencies at once, each of them will trigger `requestContentUpdate` that invokes all the renderers per call.

It is only necessary for components with multiple renderers.

Reply:
This optimization would have been unnecessary if we had had an API for triggering each renderer separately.

Reply:
What about the other question? It could be confusing if some of the renderers are synchronous and others asynchronous.

Reply:
Hmm, good question. However, `requestContentUpdate` already doesn't guarantee that the update will happen immediately, and since we are going to use that method in every directive, it turns out that they all can be updated asynchronously.

https://github.com/vaadin/web-components/blob/03dc8b8999669c7853f507745396f3b50308f1e8/packages/grid/src/vaadin-grid.js#L988

Anyway, it is a good thing to mention in the JSDoc: that the update can be performed asynchronously.

Reply:
Mentioned this detail in the JSDoc.